src/middlewared/middlewared/plugins/device_/vrrp_events.py

```diff
@@ -56,29 +56,38 @@ def run(self):
                     sleep(self._retry_timeout)
 
 
-async def _event_system(middleware, *args, **kwargs):
+async def _start_stop_vrrp_thread(middleware, *, shutting_down=False):
     global VRRP_THREAD
-    try:
-        shutting_down = args[1]['id'] == 'shutdown'
-    except (IndexError, KeyError):
-        shutting_down = False
 
     licensed = await middleware.call('failover.licensed')
-    if licensed and (VRRP_THREAD is None or not VRRP_THREAD.is_alive()):
-        # if this is a system that is being licensed for HA for the
-        # first time (without being rebooted) then we need to make
-        # sure we start this.
-        VRRP_THREAD = VrrpFifoThread(middleware=middleware)
-        VRRP_THREAD.start()
-    elif (VRRP_THREAD is not None and VRRP_THREAD.is_alive()) and not licensed or shutting_down:
+    if (not licensed or shutting_down) and (VRRP_THREAD is not None and VRRP_THREAD.is_alive()):
         # maybe the system is being downgraded to non-HA
         # (this is rare but still need to handle it) or
         # system is being restarted/shutdown etc
         await middleware.run_in_thread(VRRP_THREAD.shutdown)
         VRRP_THREAD = None
+    elif licensed and (VRRP_THREAD is None or not VRRP_THREAD.is_alive()):
+        # if this is a system that is being licensed for HA for the
+        # first time (without being rebooted) then we need to make
+        # sure we start this.
+        VRRP_THREAD = VrrpFifoThread(middleware=middleware)
+        VRRP_THREAD.start()
+
+
+async def _event_system_ready(middleware, event_type, args):
+    await _start_stop_vrrp_thread(middleware)
+
+
+async def _event_system_shutdown(middleware, event_type, args):
+    await _start_stop_vrrp_thread(middleware, shutting_down=True)
+
+
+async def _post_license_update(middleware, *args, **kwargs):
+    await _start_stop_vrrp_thread(middleware)
 
 
 async def setup(middleware):
-    middleware.create_task(_event_system(middleware))  # start thread on middlewared service start/restart
-    middleware.register_hook('system', _event_system)  # catch shutdown event and clean up thread
```

Comment:
@yocalebo this never worked, `event_subscribe` should have been used instead. Thread was managed only on middleware startup or license update. If it was sufficient, maybe we don't need to manage thread on system boot / shutdown after all?

Reply:
Probably not. The only use-cases that we need are:
1. on boot
2. on middlewared service restart
3. on license update from single to HA